components/salsa-macro-rules/src/setup_tracked_fn.rs

```diff
@@ -131,12 +131,13 @@ macro_rules! setup_tracked_fn {
             impl $Configuration {
                 fn fn_ingredient(db: &dyn $Db) -> &$zalsa::function::IngredientImpl<$Configuration> {
                     $FN_CACHE.get_or_create(db.as_dyn_database(), || {
-                        <dyn $Db as $Db>::zalsa_db(db);
+                        <dyn $Db as $Db>::register_downcaster(db);
                         db.zalsa().add_or_lookup_jar_by_type(&$Configuration)
```

Comment:
https://github.com/salsa-rs/salsa/pull/677 should allow us to merge the `register_downcaster` call into `add_or_lookup_jar_by_type` as it passes on the concrete types.